src/pools/abstract-pool.ts

```diff
@@ -0,0 +1,172 @@
+import EventEmitter from 'events'
+import type { MessageValue } from '../utility-types'
+
+export type ErrorHandler<Worker> = (this: Worker, e: Error) => void
+export type OnlineHandler<Worker> = (this: Worker) => void
+export type ExitHandler<Worker> = (this: Worker, code: number) => void
+
+export interface PoolOptions<Worker> {
+  /**
+   * A function that will listen for error event on each worker.
+   */
+  errorHandler?: ErrorHandler<Worker>
+  /**
+   * A function that will listen for online event on each worker.
+   */
+  onlineHandler?: OnlineHandler<Worker>
+  /**
+   * A function that will listen for exit event on each worker.
+   */
+  exitHandler?: ExitHandler<Worker>
+  /**
+   * This is just to avoid not useful warnings message, is used to set `maxListeners` on event emitters (workers are event emitters).
+   *
+   * @default 1000
+   */
+  maxTasks?: number
+}
+
+interface IWorker {
+  on(event: 'error', handler: ErrorHandler<this>): void
+  on(event: 'online', handler: OnlineHandler<this>): void
+  on(event: 'exit', handler: ExitHandler<this>): void
+}
+
+class PoolEmitter extends EventEmitter {}
+
+export abstract class AbstractPool<
```

Comment:
Abstract class is ok to avoid code duplication, but we should add an interface( IPool could be good as name) that only has one method to implement, the execute method.

In this way if we want to implement a pool that is not suitable to extends the Abstract class we can do it and end users still use the same method signature

Reply:
Will add this

Reply:
Exposed IPool interface in [`d00ce3f` (Define IPool interface)](https://github.com/pioardi/poolifier/pull/116/commits/d00ce3f6ae775a86eb3aeb9c645a87c2ce270575)